internal/app/orchestrator/orchestrator.go

```diff
@@ -95,16 +114,183 @@ func New(ctx context.Context,
+	// that maps to the same aggregated key doesn't result in two upstream
+	// streams.
+	o.upstreamResponseMap.mu.Lock()
+	if _, ok := o.upstreamResponseMap.responseChannels[aggregatedKey]; !ok {
+		upstreamResponseChan, shutdown, err := o.upstreamClient.OpenStream(req)
+		if err != nil {
+			// TODO implement retry/back-off logic on error scenario.
+			// https://github.com/envoyproxy/xds-relay/issues/68
+			o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "Failed to open stream to origin server")
+		} else {
+			respChannel := o.upstreamResponseMap.add(aggregatedKey, upstreamResponseChan)
+			// Spin up a go routine to watch for upstream responses.
+			// One routine is opened per aggregate key.
+			go o.watchUpstream(ctx, aggregatedKey, respChannel.response, respChannel.done, shutdown)
+		}
+
+	}
+	o.upstreamResponseMap.mu.Unlock()
+
+	return responseChannel, o.onCancelWatch(aggregatedKey, &req)
 }
 
 // Fetch implements the polling method of the config cache using a non-empty request.
-func (c *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
+func (o *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
 	return nil, fmt.Errorf("Not implemented")
 }
 
-func (c *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
-	// TODO implement.
+// watchResponse is intended to be called in a goroutine, to receive incoming
+// responses and fan out to downstream clients.
+func (o *orchestrator) watchUpstream(
+	ctx context.Context,
+	aggregatedKey string,
+	responseChannel <-chan *discovery.DiscoveryResponse,
+	done <-chan bool,
+	shutdown func(),
+) {
+	for {
+		select {
+		case x, more := <-responseChannel:
+			if !more {
+				// A problem occurred fetching the response upstream, log and
+				// return the most recent cached response, so that the
+				// downstream will reissue the discovery request.
+				o.logger.With("key", aggregatedKey).Error(ctx, "upstream error")
+			} else {
+				// Cache the response.
+				_, err := o.cache.SetResponse(aggregatedKey, *x)
+				if err != nil {
+					// If we fail to cache the new response, log and return the old one.
+					o.logger.With("err", err).With("key", aggregatedKey).
+						Error(ctx, "Failed to cache the response")
+				}
+			}
+
+			// Get downstream watches and fan out.
+			// We retrieve from cache rather than directly fanning out the
+			// newly received response because the cache does additional
+			// resource serialization.
+			cached, err := o.cache.Fetch(aggregatedKey)
+			if err != nil {
+				o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "cache fetch failed")
+				// Can't do anything because we don't know who the watches
+				// are. Drop the response.
+			} else {
+				if cached == nil || cached.Resp == nil {
+					// If cache is empty, there is nothing to fan out.
+					if !more {
+						// Warn. Benefit of the doubt that this is the first request.
```

Comment:
since this block is inside select, there will always be a response, no?

Reply:
If the channel is closed, you will continue to get the default struct and more being false.

Reply:
yeah my point was when more is false, we should return because the channel will never return further.